test/e2e/util_gatewayapi_test.go

```diff
@@ -0,0 +1,611 @@
+	// openshiftIstioOperatorDeploymentName holds the expected istio-operator deployment name.
+	openshiftIstioOperatorDeploymentName = "istio-operator"
+	// openshiftIstiodDeploymentName holds the expected istiod proxy deployment name
+	openshiftIstiodDeploymentName = "istiod-openshift-gateway"
+	// openshiftSMCPName holds the expected OSSM ServiceMeshControlPlane name
+	openshiftSMCPName = "openshift-gateway"
+)
+
+// updateIngressOperatorRole updates the ingress-operator cluster role with cluster-admin privilege.
+// TODO - Remove this function after https://issues.redhat.com/browse/OSSM-3508 is fixed.
+func updateIngressOperatorRole(t *testing.T) error {
+	t.Helper()
+
+	// Create the same rolebinding that the `oc adm policy add-cluster-role-to-user` command creates.
+	// Caller must remove this after setting it.
+	crb := &rbacv1.ClusterRoleBinding{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "cluster-admin-e2e",
+		},
+		RoleRef:  rbacv1.RoleRef{APIGroup: rbacv1.GroupName, Kind: "ClusterRole", Name: "cluster-admin"},
+		Subjects: []rbacv1.Subject{{Kind: rbacv1.ServiceAccountKind, Name: "ingress-operator", Namespace: operatorcontroller.DefaultOperatorNamespace}},
+	}
+
+	// Add the rolebinding to the ingress-operator user.
+	if err := kclient.Create(context.TODO(), crb); err != nil {
+		if kerrors.IsAlreadyExists(err) {
+			t.Logf("rolebinding already exists")
+			return nil
+		}
+		t.Logf("error adding rolebinding: %v", err)
+		return err
+	}
+	t.Log("rolebinding has been added")
+	return nil
+}
+
+// assertCrdExists checks if the CRD of the given name exists and returns an error if not.
+// Otherwise returns the CRD version.
+func assertCrdExists(t *testing.T, crdname string) (string, error) {
+	t.Helper()
+	crd := &apiextensionsv1.CustomResourceDefinition{}
+	name := types.NamespacedName{Namespace: "", Name: crdname}
+	crdVersion := ""
+
+	err := wait.PollUntilContextTimeout(context.Background(), 1*time.Second, 30*time.Second, false, func(context context.Context) (bool, error) {
+		if err := kclient.Get(context, name, crd); err != nil {
+			t.Logf("failed to get crd %s: %v", name, err)
+			return false, nil
+		}
+		crdConditions := crd.Status.Conditions
+		for _, version := range crd.Spec.Versions {
+			if version.Served {
+				crdVersion = version.Name
+			}
+		}
+		for _, c := range crdConditions {
+			if c.Type == apiextensionsv1.Established && c.Status == apiextensionsv1.ConditionTrue {
+				return true, nil
+			}
+		}
+		t.Logf("failed to find crd %s to be Established", name)
+		return false, nil
+	})
+	return crdVersion, err
+}
+
+// createHttpRoute checks if the HTTPRoute can be created.
+// If it can't an error is returned.
+func createHttpRoute(namespace, routename, parentnamespace, hostname, backendrefname string, gateway *gwapi.Gateway) (*gwapi.HTTPRoute, error) {
+	// Just in case gateway creation failed, supply a fake gateway name.
+	name := "NONE"
+	if gateway != nil {
+		name = gateway.Name
+	}
```

Comment:
*nit* I'm not totally sure about how this logic will be helpful in the case of a nil Gateway, it may be more confusing to set the parent Gateway to `NONE` and continue, rather than just throwing an error.

Maybe you should handle it like you did the in `assertHttpRouteSuccessful`:
```suggestion
        if gateway == nil {
		return nil, errors.New("unable to create httpRoute, no gateway available")
	}
```